### PR TITLE
Individual optimisers

### PR DESCRIPTION
**Pull request by tnowotny**
I have re-implemented how we designate which variables are trained with which optimiser.
In essence, during ```EventPropCompiler.compile()``` we pass an optional argument ```optimisers``` through ```**kwargs``` that is a dictionary of dictionaries, e.g.
```python
optimisers={connection: {"var": optimiser,
"var2": optimiser2},
connection2: {"var": optimiser3}
}
```
This is then passed through and interpreted so that each variable is optimised by the specified optimiser. I think ```compile()``` is the right point to specify optimisers, as one could now use the same ```EventPropCompiler``` to make a network for different training setups (different optimisers and learning rates for different parts of the network).

Furthermore, I have added training with respect to neuron parameters. The optimiser is defined through the same mechanism, e.g.
```python
optimisers={population: {"var": optimiser}}
```
The code runs simple examples, training weights, delays, and neuron parameters, including just training neuron parameters. It remains to be tested whether the gradient for training neuron parameters is correct, and this does something useful in an appropriate example.

TODO:
- [x] Implement sentinel syntax for easier specifying of defaults (I think this will allow us to remove the optimiser passed to the constructor)
- [x] Unify ``Connection``/``Layer``/``Populations`` in optimiser dictionary
- [ ] Checkpointing - should be trivial
- [ ] Optimisers should be created with ``get_object`` so you can use e.g. "adam" rather than having to import ``Adam``

### CODE REVIEW

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 1]
```diff
@@ -580,9 +574,21 @@ def pre_compile(self, network: Network,
         # Build list of output populations
         readouts = [p for p in network.populations
                     if p.neuron.readout is not None]
-
-        return CompileState(self.losses, readouts,
+        compile_state = CompileState(self.losses, readouts,
                             genn_model.backend_name)
+        # use explicit dictionary of optimisers if provided
+        if "optimisers" in kwargs.keys():
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:580]: you don't need the ``keys()``
```suggestion
        if "optimisers" in kwargs:
```

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 2]
```diff
@@ -702,18 +708,14 @@ def build_weight_update_model(self, conn: Connection,
         # Does this connection have a delay?
         has_delay = (not is_value_constant(connect_snippet.delay)
                      or connect_snippet.delay > 0)
-        
         # Does this connection have learnable delays
-        has_learnable_delay = conn in self.delay_learn_conns
+        has_learnable_delay = conn in compile_state.optimisers.keys()\
+            and "delay" in compile_state.optimisers[conn].keys()
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:713]:
```suggestion
        has_learnable_delay = (conn in compile_state.optimisers
                               and "delay" in compile_state.optimisers[conn])
```

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 3]
```diff
@@ -970,50 +972,86 @@ def create_compiled_network(self, genn_model, neuron_populations: dict,
...
-                cu_weight = self._create_optimiser_custom_update(
-                    f"Weight{i}", create_wu_var_ref(genn_pop, "weight"),
-                    create_wu_var_ref(genn_pop, "weightGradient"), 
-                    self._optimiser, genn_model)
+                gradient_vars = []
+                for var, o in optim.items():
+                    # If weight optimisation is required
+                    if var == "weight":
+                        # Create weight optimiser custom update
+                        # TODO: Should this be more general for any var in wum?
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:987]: No, WUMs aren't really a concept exposed to mlGeNN users

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 4]
```diff
@@ -1410,8 +1457,16 @@ def _build_in_hid_neuron_model(self, pop: Population,
                     "neurons defined in terms of AutoNeuronModel")
             
             # Build adjoint system from model
-            dl_dt, adjoint_jumps, saved_vars_timestep, saved_vars_spike =\
-                self._build_adjoint_system(model, False, self.reg_lambda != 0.0)
+            learn_params= [] if pop not in compile_state.optimisers \
+                else compile_state.optimisers[pop]
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:1461]: Aren't these dictionaries?
```suggestion
            learn_params = compile_state.optimisers.get(pop, {})
```

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 5]
```diff
@@ -1575,8 +1632,14 @@ def _build_out_neuron_model(self, pop: Population,
             example_timesteps=self.example_timesteps)
 
         # Build adjoint system from model
-        dl_dt, adjoint_jumps, saved_vars_timestep, saved_vars_spike =\
-            self._build_adjoint_system(model, True, False)
+        learn_params= [] if pop not in compile_state.optimisers \
+            else compile_state.optimisers[pop]
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:1636]:
```suggestion
        learn_params = compile_state.optimisers.get(pop, {})
```

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 6]
```diff
@@ -970,50 +972,86 @@ def create_compiled_network(self, genn_model, neuron_populations: dict,
...
+                    #gradient_vars.append((f"{p}Gradient", "scalar", 0.0))
+                # Create reset model for gradient variables
+                #assert len(gradient_vars) > 0
+                #zero_grad_model = create_reset_custom_update(
+                #    gradient_vars,
+                #    lambda name: create_var_ref(genn_pop, name))
+
+                # Add custom update
+                #self.add_custom_update(genn_model, zero_grad_model,
+                #                       "ZeroGradient", f"CUZeroPopGradient{i}")
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:1053]: Yes, this is correct - the normal resetting mechanism should handle this correctly for population variables

```suggestion
```

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 7]
```diff
@@ -970,50 +972,86 @@ def create_compiled_network(self, genn_model, neuron_populations: dict,
...
+                                       "ZeroGradient", f"CUZeroConnGradient{i}")
+                need_zero_gradient_update_group = True
+                i = i+1
+
+        # Loop through populations that require optimisers
+        i = 0
+        for c, optim in compile_state.optimisers.items():
+            if c in neuron_populations:
+                genn_pop = neuron_populations[c]
+                #gradient_vars = []
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:1031]:
```suggestion
```

**Review by neworderofjamie — CHANGES_REQUESTED** [review]
I'll have a go at the todo list but, in general, this looks really good
